test/functional/eval/capture_spec.lua

```diff
@@ -0,0 +1,72 @@
+local helpers = require('test.functional.helpers')
+local eq = helpers.eq
+local eval = helpers.eval
+local clear = helpers.clear
+local source = helpers.source
+local redir_exec = helpers.redir_exec
+local exc_exec = helpers.exc_exec
+local funcs = helpers.funcs
+
+describe('capture()', function()
+  before_each(clear)
+
+  it('returns the same result with :redir', function()
+    eq(redir_exec('messages'), funcs.capture('messages'))
+  end)
+
+  it('returns the output of the commands if the argument is List', function()
+    eq("foobar", funcs.capture({'echon "foo"', 'echon "bar"'}))
+    eq("\nfoo\nbar", funcs.capture({'echo "foo"', 'echo "bar"'}))
+  end)
+
+  it('supports the nested redirection', function()
+    source([[
+    function! g:Foo()
+      let a = ''
+      redir => a
+      silent echon "foo"
+      redir END
+      return a
+    endfunction
+    function! g:Bar()
+      let a = ''
+      redir => a
+      call g:Foo()
+      redir END
+      return a
+    endfunction
+    ]])
+    eq('foo', funcs.capture('call g:Bar()'))
+
+    eq('42', funcs.capture([[echon capture("echon capture('echon 42')"]]))
+  end)
+
+  it('supports the nested echo', function()
+    eq('42', funcs.capture('echon capture("echon 42")'))
+  end)
+
+  it('returns the literal string', function()
+    eq('^A', funcs.capture('echon "\\<C-a>"'))
+  end)
+
+  it('returns the empty string if the argument list is empty', function()
+    eq('', funcs.capture({}))
+    exc_exec('let g:ret = capture(v:_null_list)')
```

Comment:
If you use `exc_exec` then you need to test its output: `eq(0, exc_exec(…))`. Otherwise what you do is silencing the exceptions.